rfcs/framework-0072-pluggable-transfers.md

```diff
@@ -0,0 +1,507 @@
+- Start Date: 2024-03-22
+- RFC PR: [#<PR>](https://github.com/inveniosoftware/rfcs/pull/<PR>)
+- Authors: Mirek Simek <miroslav.simek@cesnet.cz>, Miroslav Bauer <bauer@cesnet.cz>
+
+# Pluggable Transfer Types for Record Files
+
+## Summary
+
+Currently, the system only supports three fixed transfer types—**Local**, **Fetch**,
+and **Remote**—and does not allow adding new ones. By making transfer types pluggable,
+users could implement and integrate their own custom transfers.
+
+To fully support pluggable transfer types on the front end, the file uploader UI must also be customizable.
+We propose integrating a custom file uploader UI based on the Uppy.io library, utilizing the AWS S3 Multipart uploader
+plugin.
+
+## Motivation
+
+We plan to introduce a built-in **Multipart** transfer type that enables
+uploading large files in separate chunks, improving performance and
+reliability.
+
+In addition, one could, for example, develop an **S3Copy** transfer type to
+copy files directly between S3 buckets—bypassing the Invenio server—to reduce
+bandwidth usageand improve efficiency.
+
+As a record depositor, I want to:
+- Select one or multiple local files to upload and assign to a record.
+- Select one or multiple remote files (e.g., import from a parent record or cloud storage services).
+- View the upload progress of each ongoing file upload.
+- Abort an ongoing upload.
+- Retry any failed uploads.
+- View and manage already uploaded files (e.g., remove files, update associated metadata such as image captions and descriptions).
+- Verify file integrity using checksums when uploading (to ensure the file was received and stored without errors) or downloading (to confirm I received an exact copy as stored on the server).
+- Efficiently and reliably upload large files (>10 GiB) even over unreliable networks.
+
+As a developer, I want to:
+- Implement or integrate custom file uploader UIs (including custom upload logic) into the deposit form (e.g., using library like Uppy.io), allowing the front-end to support any configured transfer types.
+
+As an admin, I want to:
+- Configure the default upload mechanism (TransferType) for new uploads.
+- Configure which uploader UI should be used in the deposit form.
```

Comment:
Thank you for very clear use cases, it reads well!

I am curious, what would be a motivation behind an admin wanting to change the UI of the file uploader? Do you mean in application runtime?

Reply:
Thanks a lot for your review. This point comes from a discussion here: https://github.com/inveniosoftware/rfcs/issues/86#issuecomment-2575240491. I had an impression, that it would be useful for the admin to be able to choose wherever to keep the current simple UI, use the more powerful Uppy uploader, or to choose any other custom uploader UI implementation for a site. This is configured at build time.

Reply:
I see :) I was writing this with "admin" meaning actually the devops guy, who configures `invenio.cfg` / runtime env config for a site.